--- BOSS_SM.cpp	2020-08-10 08:22:35.729797000 +0000
+++ BOSS_SM.cpp.patched	2020-08-10 08:23:48.764000000 +0000
@@ -3,71 +3,71 @@
 #include "/root/gambit_THDM/Backends/scripts/BOSS/modules/../../../../Backends/installed/THDMC/1.8.0/src/SM.h"
 
 
-const double& SM::alpha_ref__BOSS() { return alpha; }
+const double& SM::alpha_ref__BOSS() { return double(alpha); }
 
-const double& SM::alpha0_ref__BOSS() { return alpha0; }
+const double& SM::alpha0_ref__BOSS() { return double(alpha0); }
 
-const double& SM::GF_ref__BOSS() { return GF; }
+const double& SM::GF_ref__BOSS() { return double(GF); }
 
-const double& SM::MZ_ref__BOSS() { return MZ; }
+const double& SM::MZ_ref__BOSS() { return double(MZ); }
 
-const double& SM::MW_ref__BOSS() { return MW; }
+const double& SM::MW_ref__BOSS() { return double(MW); }
 
-const double& SM::alpha_s_ref__BOSS() { return alpha_s; }
+const double& SM::alpha_s_ref__BOSS() { return double(alpha_s); }
 
-const double& SM::GammaZ_ref__BOSS() { return GammaZ; }
+const double& SM::GammaZ_ref__BOSS() { return double(GammaZ); }
 
-const double& SM::GammaW_ref__BOSS() { return GammaW; }
+const double& SM::GammaW_ref__BOSS() { return double(GammaW); }
 
-const double& SM::md_p_ref__BOSS() { return md_p; }
+const double& SM::md_p_ref__BOSS() { return double(md_p); }
 
-const double& SM::mu_p_ref__BOSS() { return mu_p; }
+const double& SM::mu_p_ref__BOSS() { return double(mu_p); }
 
-const double& SM::ms_p_ref__BOSS() { return ms_p; }
+const double& SM::ms_p_ref__BOSS() { return double(ms_p); }
 
-const double& SM::Q_ms_ref__BOSS() { return Q_ms; }
+const double& SM::Q_ms_ref__BOSS() { return double(Q_ms); }
 
-const double& SM::mc_p_ref__BOSS() { return mc_p; }
+const double& SM::mc_p_ref__BOSS() { return double(mc_p); }
 
-const double& SM::mb_p_ref__BOSS() { return mb_p; }
+const double& SM::mb_p_ref__BOSS() { return double(mb_p); }
 
-const double& SM::mt_p_ref__BOSS() { return mt_p; }
+const double& SM::mt_p_ref__BOSS() { return double(mt_p); }
 
-const double& SM::me_p_ref__BOSS() { return me_p; }
+const double& SM::me_p_ref__BOSS() { return double(me_p); }
 
-const double& SM::mmu_p_ref__BOSS() { return mmu_p; }
+const double& SM::mmu_p_ref__BOSS() { return double(mmu_p); }
 
-const double& SM::mtau_p_ref__BOSS() { return mtau_p; }
+const double& SM::mtau_p_ref__BOSS() { return double(mtau_p); }
 
-const double& SM::Vud_ref__BOSS() { return Vud; }
+const double& SM::Vud_ref__BOSS() { return double(Vud); }
 
-const double& SM::Vus_ref__BOSS() { return Vus; }
+const double& SM::Vus_ref__BOSS() { return double(Vus); }
 
-const double& SM::Vub_ref__BOSS() { return Vub; }
+const double& SM::Vub_ref__BOSS() { return double(Vub); }
 
-const double& SM::Vcd_ref__BOSS() { return Vcd; }
+const double& SM::Vcd_ref__BOSS() { return double(Vcd); }
 
-const double& SM::Vcs_ref__BOSS() { return Vcs; }
+const double& SM::Vcs_ref__BOSS() { return double(Vcs); }
 
-const double& SM::Vcb_ref__BOSS() { return Vcb; }
+const double& SM::Vcb_ref__BOSS() { return double(Vcb); }
 
-const double& SM::Vtd_ref__BOSS() { return Vtd; }
+const double& SM::Vtd_ref__BOSS() { return double(Vtd); }
 
-const double& SM::Vts_ref__BOSS() { return Vts; }
+const double& SM::Vts_ref__BOSS() { return double(Vts); }
 
-const double& SM::Vtb_ref__BOSS() { return Vtb; }
+const double& SM::Vtb_ref__BOSS() { return double(Vtb); }
 
-const double& SM::Q_HD_ref__BOSS() { return Q_HD; }
+const double& SM::Q_HD_ref__BOSS() { return double(Q_HD); }
 
-const bool& SM::b_HD_ref__BOSS() { return b_HD; }
+const bool& SM::b_HD_ref__BOSS() { return bool(b_HD); }
 
-const double& SM::ms_5_ref__BOSS() { return ms_5; }
+const double& SM::ms_5_ref__BOSS() { return double(ms_5); }
 
-const double& SM::mc_5_ref__BOSS() { return mc_5; }
+const double& SM::mc_5_ref__BOSS() { return double(mc_5); }
 
-const double& SM::mb_5_ref__BOSS() { return mb_5; }
+const double& SM::mb_5_ref__BOSS() { return double(mb_5); }
 
-const double& SM::mt_5_ref__BOSS() { return mt_5; }
+const double& SM::mt_5_ref__BOSS() { return double(mt_5); }
 
 
 #include "backend_types/THDMC_1_8_0/identification.hpp"
